query-algebrizer/src/clauses/where_fn.rs

```diff
@@ -0,0 +1,384 @@
+            FnArg::Constant(NonIntegerConstant::Text(x)) => Ok(TypedValue::String(x)),
+            FnArg::Variable(var) => lookup(var),
+            FnArg::Vector(_) |
+            FnArg::SrcVar(_) => bail!(ErrorKind::InvalidGroundConstant),
+        }
+    }
+
+    fn into_column_vector(self, lookup: F) -> Result<Vec<TypedValue>> {
+        if let FnArg::Vector(children) = self {
+            children.into_iter().map(|x| x.into_typed_value(&lookup)).collect()
+        } else {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+    }
+
+    /// Convert an atom into a 1x1 matrix.
+    fn into_unit_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_typed_value(lookup).map(|x| vec![vec![x]])
+    }
+
+    /// Convert a vector of atoms into a 1xN matrix.
+    fn into_row_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_column_vector(lookup).map(|x| vec![x])
+    }
+
+    /// Convert a vector of atoms into an Mx1 matrix.
+    fn into_column_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        self.into_column_vector(lookup).map(|x| x.into_iter().map(|y| vec![y]).collect())
+    }
+
+    /// Convert a vector of vectors (each of atoms) into an MxN matrix.
+    fn into_block_matrix(self, lookup: F) -> Result<Vec<Vec<TypedValue>>> {
+        if let FnArg::Vector(children) = self {
+            children.into_iter().map(|x| x.into_column_vector(&lookup)).collect()
+        } else {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+    }
+}
+
+/// This function takes a vector of rows of possibly different lengths and a sequence of columns to
+/// ignore.  It ensures that each row has the same length, and drops the columns that have `false`
+/// in the `column_is_bound` sequence.  Finally, it ensures that each column is homogeneously typed,
+/// and returns a sequence of the derived type for each column.
+fn reduced_rows_and_column_types(rows: Vec<Vec<TypedValue>>, column_is_bound: &Vec<bool>) -> Result<(Vec<Vec<TypedValue>>, Vec<ValueType>)> {
+    let mut matrix = Vec::with_capacity(rows.len());
+
+    for row in rows {
+        if column_is_bound.len() != row.len() {
+            bail!(ErrorKind::InvalidGroundConstant)
+        }
+
+        let mut retained_row = Vec::with_capacity(column_is_bound.len());
+        for (is_bound, value) in column_is_bound.into_iter().zip(row.into_iter()) {
+            if *is_bound {
+                retained_row.push(value);
+            }
+        }
+
+        matrix.push(retained_row);
+    }
+
+    let column_types: Vec<ValueType> = match matrix.get(0) {
+        None => {
+            bail!(ErrorKind::InvalidGroundConstant)
+        },
+        Some(first_row) => {
+            first_row.iter().map(|x| x.value_type()).collect()
+        },
+    };
+
+    for row in matrix.iter() {
+        for (column_type, value) in column_types.iter().zip(row.into_iter()) {
+            if *column_type != value.value_type() {
```

Comment:
If you maintain a vector of type sets, you can broaden as you go during the initial walk on line 110.

When you get to here you can throw if you wish if any of them are non-unit, or you can implement tag projection.

Preserving type sets means you can do something useful for:

```
[:find …
 :where
 [(ground [[10 :foo] ["string" :bar]] [?x ?y]]     ;; ?x could be ref, long, or string.
 [?x :foo/keyword ?y]]]                            ;; ?x can only be ref!
```

There is a little bit of trickiness here, because `ground` can introduce multiple sets of bindings, and not all of them can be congruent with the rest of the query! Those can be pruned at this point. But the type codes must be projected for ambiguous columns.

Note that you can't quite get away from handling this ambiguity. If the input is EDN, `10` could be `TypedValue::Ref(10)` or `TypedValue::Long(10)`. A `ground` implementation that doesn't project the type tag, and doesn't join that tag against the rest of the query, will produce incorrect results.